File: solutions/108. Convert Sorted Array to Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        int lo = 0;
        int hi = nums.length - 1;
​
        TreeNode root = this.sortedArrayToBST(nums, lo, hi);
        return root;
    }
​
    private TreeNode sortedArrayToBST(int[] nums, int lo, int hi) {
        if (lo > hi) return null;
        
        if (lo == hi) return new TreeNode(nums[lo]);
​
        int mid = (lo + hi) / 2;
        TreeNode node = new TreeNode(nums[mid]);
​
        // faith
        node.left = this.sortedArrayToBST(nums, lo, mid - 1);
        node.right = this.sortedArrayToBST(nums, mid + 1, hi);
​
        return node;
    }
}

File: ..solutions/108. Convert Sorted Array to Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    TreeNode* f(vector<int> &nums, int lo, int hi){
        if(lo > hi){
            return nullptr;
        }
        int mid = (lo + hi) >> 1;
        TreeNode* root = new TreeNode(nums[mid]);
        root->left = f(nums, lo, mid-1);
        root->right = f(nums, mid + 1, hi);
        return root;
    }
    
    TreeNode* sortedArrayToBST(vector<int>& nums) {
       if(nums.empty()){
           return nullptr;
       }
        return f(nums, 0, nums.size() - 1);
    }
};
